klib/ntp.c

```diff
@@ -0,0 +1,158 @@
+    ntptime->fraction = PP_HTONL((u32)t);
+}
+
+static timestamp ntptime_to_timestamp(struct ntp_ts *ntptime)
+{
+    return (seconds(PP_NTOHL(ntptime->seconds) - NTP_EPOCH_DELTA) + PP_NTOHL(ntptime->fraction));
+}
+
+static s64 ntptime_diff(struct ntp_ts *t1, struct ntp_ts *t2)
+{
+    return (seconds(PP_NTOHL(t1->seconds) - PP_NTOHL(t2->seconds)) +
+            PP_NTOHL(t1->fraction) - PP_NTOHL(t2->fraction));
+}
+
+static void ntp_query(const ip_addr_t *server_addr)
+{
+    struct pbuf *p = ntp.pbuf_alloc(PBUF_TRANSPORT, sizeof(struct ntp_packet), PBUF_RAM);
+    if (p == 0)
+        return;
+    struct ntp_packet *pkt = p->payload;
+    ntp.runtime_memset(p->payload, 0, sizeof(*pkt));
+    pkt->vn = 3;    /* NTP version number */
+    pkt->mode = 3;  /* client mode */
+    timestamp_to_ntptime(ntp.now(CLOCK_ID_REALTIME), &pkt->transmit_ts);
+    ntp.udp_sendto(ntp.pcb, p, server_addr, NTP_PORT);
+    ntp.pbuf_free(p);
+}
+
+static void ntp_input(void *z, struct udp_pcb *pcb, struct pbuf *p,
+                      const ip_addr_t *addr, u16 port)
+{
+    struct ntp_packet *pkt = p->payload;
+    if (p->len == sizeof(*pkt)) {
+        timestamp wallclock_now = ntp.now(CLOCK_ID_REALTIME);
+        timestamp origin = ntptime_to_timestamp(&pkt->originate_ts);
+        /* round trip delay */
+        timestamp rtd = wallclock_now - origin - ntptime_diff(&pkt->transmit_ts, &pkt->receive_ts);
+        s64 offset = ntptime_to_timestamp(&pkt->transmit_ts) - wallclock_now + rtd / 2;
+        double temp_cal, cal;
+        timestamp raw = ntp.now(CLOCK_ID_MONOTONIC_RAW);
+
+        /* Apply maximum slew rate until local time is synchronized with NTP time. */
+        timestamp sync_complete;
+        if (offset == 0) {
+            temp_cal = 0;
+            sync_complete = raw;
+        } else {
+            temp_cal = (offset > 0) ? NTP_MAX_SLEW_RATE : -NTP_MAX_SLEW_RATE;
+            sync_complete = raw + offset / temp_cal;
+        }
+
+        /* If at least 2 samples have been received from NTP server, calculate calibration value to
+         * to be applied after local time is synchronized with NTP time. */
+        if ((ntp.last_raw != 0) && (raw != ntp.last_raw)) {
+            cal = (double)offset / (raw - ntp.last_raw);
+            if (cal > NTP_MAX_SLEW_RATE)
+                cal = NTP_MAX_SLEW_RATE;
+            else if (cal < -NTP_MAX_SLEW_RATE)
+                cal = -NTP_MAX_SLEW_RATE;
+        } else {
+            cal = 0;
+        }
+        ntp.last_raw = raw;
+
+        ntp.clock_adjust(wallclock_now + offset, temp_cal, sync_complete, cal);
+    }
+    ntp.pbuf_free(p);
+}
+
+static void ntp_dns_cb(const char *name, const ip_addr_t *ipaddr, void *callback_arg)
+{
+    if (ipaddr)
+        ntp_query(ipaddr);
+}
+
+define_closure_function(0, 1, void, ntp_query_func,
+                        u64, overruns)
+{
+    ip_addr_t server_addr;
+    err_t err = ntp.dns_gethostbyname(NTP_SERVER, &server_addr, ntp_dns_cb, 0);
```

Comment:
Perhaps the ntp server address and port could be configurable through the manifest.

Reply:
The NTP server address and port are now configurable through the manifest, via entries such as below:
```
    ntp_address:pool.ntp.org
    ntp_port:123
```